### ISSUE TITLE
InternLM2-7b-chat微调时报`epoch_length` must be a positive integer, but got 0.

### ISSUE THREAD

**Issue by Egber1t** [InternLM/xtuner#336]
![image](https://github.com/InternLM/xtuner/assets/137043350/4af693d9-21ea-4ab1-a5f3-c7f5e329b978)
# Model
pretrained_model_name_or_path = '/home/mice/TNB/internlm2-chat-7b'

# Data
data_path = '/home/mice/TNB/test_with_system_first.jsonl'
prompt_template = PROMPT_TEMPLATE.default
max_length = 2048
pack_to_max_length = True

# Scheduler & Optimizer
batch_size = 1  # per_device
accumulative_counts = 16
dataloader_num_workers = 0
max_epochs = 1
optim_type = AdamW
lr = 2e-4
betas = (0.9, 0.999)
weight_decay = 0
max_norm = 1  # grad clip
warmup_ratio = 0.03

**Comment by pppppM** [comment]
xtuner 默认是会拼接多条数据至 2048 tokens 长度的，可能是因为数据集太短，拼接后数据条数太少导致的

可以尝试将 config 中的 pack_to_max_length 改为 False

另外，InternLM2 需要 xtuner v0.1.12 版本，看你截图里好像用的 v0.1.9

**Comment by Upcreat** [comment]
我也会出现这个问题

**Comment by LZHgrla** [comment]
@Egber1t @Upcreat  请张贴一下config中第三部分的定义信息，以及自定义数据集格式～
```
#######################################################################
#                      PART 3  Dataset & Dataloader                   #
#######################################################################
....
```

**Comment by Upcreat** [comment]
![image](https://github.com/InternLM/xtuner/assets/83943594/fd8c6a19-8db5-45d9-9704-79101fdd176e)
![image](https://github.com/InternLM/xtuner/assets/83943594/ce08ed1a-9615-4e94-a687-166706f0b0ba)

**Comment by LZHgrla** [comment]
@Upcreat  第95行不应被注释，可以加上试一下

**Comment by Upcreat** [comment]
> @Upcreat 第95行不应被注释，可以加上试一下

好的，我试下

**Comment by Egber1t** [comment]
@LZHgrla 麻烦您看一下，谢谢！
![image](https://github.com/InternLM/xtuner/assets/137043350/fb7473e6-f600-4ab9-afdd-0c4e2996ddcf)
![image](https://github.com/InternLM/xtuner/assets/137043350/f77962a2-3072-4249-8de6-4126dabddba0)
![image](https://github.com/InternLM/xtuner/assets/137043350/bc6ec8e7-1655-47a6-a17b-a1d3460e2bff)
![image](https://github.com/InternLM/xtuner/assets/137043350/04232335-8100-45d7-8139-d5f135c1e3d8)
我是拿最后一个jsonl导入到data路径下

**Comment by Upcreat** [comment]
> @Upcreat 第95行不应被注释，可以加上试一下

![image](https://github.com/InternLM/xtuner/assets/83943594/97cc03e3-7abf-43e4-8017-f487382f29e1)
我的现在可以正常训练了，看下等下什么情况。感谢您！！！

**Comment by LZHgrla** [comment]
@Egber1t 需要针对这个自定义数据集的“字段”，写好自定义的map_fn，具体可以参考这个demo https://github.com/InternLM/xtuner/tree/main/examples/demo_data/single_turn

**Comment by Egber1t** [comment]
> @Egber1t 需要针对这个自定义数据集的“字段”，写好自定义的map_fn，具体可以参考这个demo https://github.com/InternLM/xtuner/tree/main/examples/demo_data/single_turn

已解决，谢谢大佬，我每个加上了一个"conversation":直接放入json，就可以了。然后我想问一下，出来的lora权重合并之后，我怎么去进行评测呢，比如bleu指标等

**Comment by Upcreat** [comment]
> @Upcreat 第95行不应被注释，可以加上试一下

我这边可以正常训练了，感谢！！！

**Comment by LZHgrla** [comment]
@Egber1t  在训练前会打印数据信息，可以再确认一下是否准确load进来了所有数据。

评测的话建议使用 opencompass，其已经支持了 InternLM2-Chat 的评测，同时有众多数据集可以自由组合

https://opencompass.readthedocs.io/zh-cn/latest/index.html

**Comment by Egber1t** [comment]
> @Egber1t 手动加"conversation"的话需要调整成这种格式 [https://github.com/InternLM/xtuner/blob/main/docs/zh_cn/user_guides/dataset_format.md#单轮对话数据集格式](https://github.com/InternLM/xtuner/blob/main/docs/zh_cn/user_guides/dataset_format.md#%E5%8D%95%E8%BD%AE%E5%AF%B9%E8%AF%9D%E6%95%B0%E6%8D%AE%E9%9B%86%E6%A0%BC%E5%BC%8F)
> 
> 可以查看一下是否修改成了这个样子。
> 
> 在训练前会打印数据信息，可以再确认一下是否准确load进来了所有数据。
> 
> 评测的话建议使用 opencompass，其已经支持了 InternLM2-Chat 的评测，同时有众多数据集可以自由组合
> 
> https://opencompass.readthedocs.io/zh-cn/latest/index.html

已经微调成功，谢谢LZH大佬的帮助！

**Comment by Egber1t** [comment]
@LZHgrla https://github.com/InternLM/xtuner/issues/337，佬再帮忙解决一下issue，谢谢！
